Source/WebCore/css/CSSStyleSheet.cpp

```diff
@@ -367,9 +367,14 @@ ExceptionOr<unsigned> CSSStyleSheet::insertRule(const String& ruleString, unsign
 
     RuleMutationScope mutationScope(this, RuleInsertion, dynamicDowncast<StyleRuleKeyframes>(*rule));
 
+    bool isNamespace = rule->isNamespaceRule();
     bool success = m_contents.get().wrapperInsertRule(rule.releaseNonNull(), index);
-    if (!success)
+    if (!success) {
+        if (isNamespace)
+            return Exception { ExceptionCode::InvalidStateError };
         return Exception { ExceptionCode::HierarchyRequestError };
+    }
```

Comment:
isNamespace does not need to be its own variable.

This might also work:

```suggestion
    if (!success)
        return Exception { rule->isNamespaceRule() ? ExceptionCode::InvalidStateError : ExceptionCode::HierarchyRequestError };
```

Reply:
It leads to crash on test case.

```
Thread 0 Crashed:
0   WebCore                       	       0x11808e47c WebCore::CSSStyleSheet::insertRule(WTF::String const&, unsigned int) + 560
1   WebCore                       	       0x116f29090 WebCore::jsCSSStyleSheetPrototypeFunction_insertRuleBody(JSC::JSGlobalObject*, JSC::CallFrame*, WebCore::JSCSSStyleSheet*) + 212
2   ???                           	       0x124acc044 ???
3   ???                           	       0x124ac0068 ???
4   ???                           	       0x124ac0068 ???
5   ???                           	       0x124ac0068 ???
6   ???                           	       0x124ac0248 ???
7   ???                           	       0x124ac0068 ???
8   ???                           	       0x124ac0248 ???
9   ???                           	       0x124ac0008 ???
10  ???                           	       0x124ac0008 ???
11  ???                           	       0x124ac0548 ???
12  JavaScriptCore                	       0x10ecdafb8 JSC::Interpreter::executeProgram(JSC::SourceCode const&, JSC::JSGlobalObject*, JSC::JSObject*) + 992
13  JavaScriptCore                	       0x10eeda78c JSC::evaluate(JSC::JSGlobalObject*, JSC::SourceCode const&, JSC::JSValue, WTF::NakedPtr<JSC::Exception>&) + 212
14  WebCore                       	       0x117f2b35c WebCore::ScriptController::evaluateInWorld(WebCore::ScriptSourceCode const&, WebCore::DOMWrapperWorld&) + 372
15  WebCore                       	       0x117f2b870 WebCore::ScriptController::evaluateIgnoringException(WebCore::ScriptSourceCode const&) + 52
16  WebCore                       	       0x11835e2dc WebCore::ScriptElement::executeClassicScript(WebCore::ScriptSourceCode const&) + 248
17  WebCore                       	       0x11835c7a4 WebCore::ScriptElement::prepareScript(WTF::TextPosition const&) + 1544
18  WebCore                       	       0x1186983d0 WebCore::HTMLScriptRunner::runScript(WebCore::ScriptElement&, WTF::TextPosition const&) + 148
19  WebCore                       	       0x1186982b4 WebCore::HTMLScriptRunner::execute(WTF::Ref<WebCore::ScriptElement, WTF::RawPtrTraits<WebCore::ScriptElement>, WTF::DefaultRefDerefTraits<WebCore::ScriptElement>>&&, WTF::TextPosition const&) + 100
20  WebCore                       	       0x11867dcbc WebCore::HTMLDocumentParser::runScriptsForPausedTreeBuilder() + 104
21  WebCore                       	       0x11867e1b0 WebCore::HTMLDocumentParser::pumpTokenizerLoop(WebCore::HTMLDocumentParser::SynchronousMode, bool, WebCore::PumpSession&) + 824
22  WebCore                       	       0x11867d8a8 WebCore::HTMLDocumentParser::pumpTokenizer(WebCore::HTMLDocumentParser::SynchronousMode) + 108
23  WebCore                       	       0x11867ebe0 WebCore::HTMLDocumentParser::notifyFinished(WebCore::PendingScript&) + 144
24  WebCore                       	       0x11834887c WebCore::PendingScript::notifyFinished(WebCore::LoadableScript&) + 80
25  WebCore                       	       0x118326754 WebCore::LoadableScript::notifyClientFinished() + 536
26  WebCore                       	       0x1183260a4 WebCore::LoadableNonModuleScriptBase::notifyFinished(WebCore::CachedResource&, WebCore::NetworkLoadMetrics const&, WebCore::LoadWillContinueInAnotherProcess) + 188
27  WebCore                       	       0x1188e74a0 WebCore::CachedResource::checkNotify(WebCore::NetworkLoadMetrics const&, WebCore::LoadWillContinueInAnotherProcess) + 156
28  WebCore                       	       0x1188e2778 WebCore::CachedResource::finishLoading(WebCore::FragmentedSharedBuffer const*, WebCore::NetworkLoadMetrics const&) + 120
29  WebCore                       	       0x1188fc380 WebCore::CachedScript::finishLoading(WebCore::FragmentedSharedBuffer const*, WebCore::NetworkLoadMetrics const&) + 100
30  WebCore                       	       0x1188ba910 WebCore::SubresourceLoader::didFinishLoading(WebCore::NetworkLoadMetrics const&) + 692
31  WebKit                        	       0x108dfa478 WebKit::WebResourceLoader::didFinishResourceLoad(WebCore::NetworkLoadMetrics&&) + 244 (WebResourceLoader.cpp:289)
32  WebKit                        	       0x1085c5b9c auto void IPC::callMemberFunction<WebKit::WebResourceLoader, WebKit::WebResourceLoader, void (WebCore::NetworkLoadMetrics&&), std::__1::tuple<WebCore::NetworkLoadMetrics>>(WebKit::WebResourceLoader*, void (WebKit::WebResourceLoader::*)(WebCore::NetworkLoadMetrics&&), std::__1::tuple<WebCore::NetworkLoadMetrics>&&)::'lambda'(auto&&...)::operator()<WebCore::NetworkLoadMetrics>(auto&&...) const + 68 (HandleMessage.h:135) [inlined]
33  WebKit                        	       0x1085c5b9c decltype(std::declval<WebKit::WebResourceLoader>()(std::declval<WebCore::NetworkLoadMetrics>())) std::__1::__invoke[abi:sn200100]<void IPC::callMemberFunction<WebKit::WebResourceLoader, WebKit::WebResourceLoader, void (WebCore::NetworkLoadMetrics&&), std::__1::tuple<WebCore::NetworkLoadMetrics>>(WebKit::WebResourceLoader*, void (WebKit::WebResourceLoader::*)(WebCore::NetworkLoadMetrics&&), std::__1::tuple<WebCore::NetworkLoadMetrics>&&)::'lambda'(auto&&...), WebCore::NetworkLoadMetrics>(WebKit::WebResourceLoader&&, WebCore::NetworkLoadMetrics&&) + 68 (invoke.h:179) [inlined]
34  WebKit                        	       0x1085c5b9c decltype(auto) std::__1::__apply_tuple_impl[abi:sn200100]<void IPC::callMemberFunction<WebKit::WebResourceLoader, WebKit::WebResourceLoader, void (WebCore::NetworkLoadMetrics&&), std::__1::tuple<WebCore::NetworkLoadMetrics>>(WebKit::WebResourceLoader*, void (WebKit::WebResourceLoader::*)(WebCore::NetworkLoadMetrics&&), std::__1::tuple<WebCore::NetworkLoadMetrics>&&)::'lambda'(auto&&...), std::__1::tuple<WebCore::NetworkLoadMetrics>, 0ul>(WebKit::WebResourceLoader&&, WebKit::WebResourceLoader&&, std::__1::__tuple_indices<0ul>) + 68 (tuple:1375) [inlined]
35  WebKit                        	       0x1085c5b9c decltype(auto) std::__1::apply[abi:sn200100]<void IPC::callMemberFunction<WebKit::WebResourceLoader, WebKit::WebResourceLoader, void (WebCore::NetworkLoadMetrics&&), std::__1::tuple<WebCore::NetworkLoadMetrics>>(WebKit::WebResourceLoader*, void (WebKit::WebResourceLoader::*)(WebCore::NetworkLoadMetrics&&), std::__1::tuple<WebCore::NetworkLoadMetrics>&&)::'lambda'(auto&&...), std::__1::tuple<WebCore::NetworkLoadMetrics>>(WebKit::WebResourceLoader&&, WebKit::WebResourceLoader&&) + 68 (tuple:1379) [inlined]
36  WebKit                        	       0x1085c5b9c void IPC::callMemberFunction<WebKit::WebResourceLoader, WebKit::WebResourceLoader, void (WebCore::NetworkLoadMetrics&&), std::__1::tuple<WebCore::NetworkLoadMetrics>>(WebKit::WebResourceLoader*, void (WebKit::WebResourceLoader::*)(WebCore::NetworkLoadMetrics&&), std::__1::tuple<WebCore::NetworkLoadMetrics>&&) + 68 (HandleMessage.h:132) [inlined]
37  WebKit                        	       0x1085c5b9c void IPC::handleMessage<Messages::WebResourceLoader::DidFinishResourceLoad, IPC::Connection, WebKit::WebResourceLoader, WebKit::WebResourceLoader, void (WebCore::NetworkLoadMetrics&&)>(IPC::Connection&, IPC::Decoder&, WebKit::WebResourceLoader*, void (WebKit::WebResourceLoader::*)(WebCore::NetworkLoadMetrics&&)) + 128 (HandleMessage.h:337)
38  WebKit                        	       0x1085c54f8 WebKit::WebResourceLoader::didReceiveMessage(IPC::Connection&, IPC::Decoder&) + 164 (WebResourceLoaderMessageReceiver.cpp:78)
39  WebKit                        	       0x108df1ee4 WebKit::NetworkProcessConnection::dispatchMessage(IPC::Connection&, IPC::Decoder&) + 188 (NetworkProcessConnection.cpp:103)
40  WebKit                        	       0x108446e5c WebKit::NetworkProcessConnection::didReceiveMessage(IPC::Connection&, IPC::Decoder&) + 76 (NetworkProcessConnectionMessageReceiver.cpp:99)
41  WebKit                        	       0x10916dfa0 IPC::Connection::dispatchMessage(IPC::Decoder&) + 176 (Connection.cpp:1383)
42  WebKit                        	       0x10916e13c IPC::Connection::dispatchMessage(WTF::UniqueRef<IPC::Decoder>) + 152 (Connection.cpp:1433)
43  WebKit                        	       0x10916e2e8 IPC::Connection::dispatchOneIncomingMessage() + 128 (Connection.cpp:1512)
44  JavaScriptCore                	       0x10e245b64 WTF::RunLoop::performWork() + 552
45  JavaScriptCore                	       0x10e2466e4 WTF::RunLoop::performWork(void*) + 40
46  CoreFoundation                	       0x183293bc8 __CFRUNLOOP_IS_CALLING_OUT_TO_A_SOURCE0_PERFORM_FUNCTION__ + 28
47  CoreFoundation                	       0x183293b5c __CFRunLoopDoSource0 + 172
48  CoreFoundation                	       0x1832938c8 __CFRunLoopDoSources0 + 232
49  CoreFoundation                	       0x183292558 __CFRunLoopRun + 820
50  CoreFoundation                	       0x183350988 _CFRunLoopRunSpecificWithOptions + 532
51  Foundation                    	       0x1854bea70 -[NSRunLoop(NSRunLoop) runMode:beforeDate:] + 212
52  Foundation                    	       0x184ad29b0 -[NSRunLoop(NSRunLoop) run] + 64
53  libxpc.dylib                  	       0x182ee7e30 _xpc_objc_main + 700
54  libxpc.dylib                  	       0x182ef8cf8 _xpc_main + 40
55  libxpc.dylib                  	       0x182ee79d0 xpc_main + 64
56  WebKit                        	       0x1084c735c WebKit::XPCServiceMain(int, char const**) + 44 (XPCServiceMain.mm:296)
57  dyld                          	       0x182e35d54 start + 7184
```

Reply:
How about:
```
    if (!success)
        return rule->isNamespaceRule() ? Exception { ExceptionCode::InvalidStateError } : Exception { ExceptionCode::HierarchyRequestError };
```

Reply:
@nt1m - Still crashes.

Reply:
Oh the crash comes from `.releaseNonNull()`

Reply:
Yep.